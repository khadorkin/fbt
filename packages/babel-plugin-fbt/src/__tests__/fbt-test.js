/**
 * Copyright 2004-present Facebook. All Rights Reserved.
 *
 * @emails oncall+i18n_fbt_js
 * @format
 */

jest.autoMockOff();

const {
  jsCodeFbtCallSerializer,
  snapshotTransform,
  withFbtRequireStatement,
} = require('../FbtTestUtil');

expect.addSnapshotSerializer(jsCodeFbtCallSerializer);

function runTest(data, extra) {
  expect(snapshotTransform(data.input, extra)).toMatchSnapshot();
}

// TODO(T40113359) Re-enable once this test scenario is ready to be tested
// eslint-disable-next-line jest/no-disabled-tests
xdescribe('fbt() API: ', () => {
  describe('using extraOptions', () => {
    it('should accept "locale" extra option', () => {
      runTest(
        {
          input: withFbtRequireStatement(
            `fbt("Foo", "Bar", {locale: locale.data});`,
          ),
        },
        {
          extraOptions: {locale: true},
        },
      );
    });
  });

  describe('using FBT subject', () => {
    it('should accept "subject" as a parameter', () => {
      runTest({
        input: withFbtRequireStatement(`fbt("Foo", "Bar", {subject: foo});`),
      });
    });
  });

  describe('using FBT subject with string templates', () => {
    it('should accept "subject" as a parameter', () => {
      runTest({
        input: withFbtRequireStatement('fbt(`Foo`, "Bar", {subject: foo});'),
      });
    });
  });
});

// TODO(T40113359) Re-enable once this test scenario is ready to be tested
// eslint-disable-next-line jest/no-disabled-tests
xdescribe('Test double-lined params', () => {
  it('should remove the new line for param names that are two lines', () => {
    runTest({
      input: withFbtRequireStatement(
        `<fbt desc="d">
          <fbt:param
            name="two
                  lines">
            <b>
              <fbt desc="test">simple</fbt>
            </b>
          </fbt:param>
          test
        </fbt>`,
      ),
    });
  });
});

// TODO(T40113359) Re-enable once this test scenario is ready to be tested
// eslint-disable-next-line jest/no-disabled-tests
xdescribe('fbt variable binding detection', () => {
  function describeFbtBindingTestCase(requireStatement) {
    return {
      input: `${requireStatement};
        fbt("Foo", "Bar");`,
    };
  }

  it(`should handle commonJS require()`, () => {
    runTest(describeFbtBindingTestCase(`const fbt = require('fbt')`));
  });

  describe('using ES6', () => {
    it(`should handle fbt default export`, () => {
      runTest(describeFbtBindingTestCase(`import fbt from 'fbt'`));
    });
    it(`should handle the named fbt export`, () => {
      runTest(describeFbtBindingTestCase(`import {fbt} from 'fbt'`));
    });
  });
});
