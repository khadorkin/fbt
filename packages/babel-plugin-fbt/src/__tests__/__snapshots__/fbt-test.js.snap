// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test double-lined params should remove the new line for param names that are two lines 1`] = `
const fbt = require("fbt");

fbt._(
  /* __FBT__ start */ {
    type: "text",
    jsfbt: "{two lines} test",
    desc: "d",
    project: "",
  } /* __FBT__ end */,
  [
    fbt._param(
      "two lines",
      /*#__PURE__*/ React.createElement(
        "b",
        null,
        fbt._(
          /* __FBT__ start */ {
            type: "text",
            jsfbt: "simple",
            desc: "test",
            project: "",
          } /* __FBT__ end */
        )
      )
    ),
  ]
);

`;

exports[`fbt variable binding detection should handle commonJS require() 1`] = `
const fbt = require("fbt");

fbt._(
  /* __FBT__ start */ {
    type: "text",
    jsfbt: "Foo",
    desc: "Bar",
    project: "",
  } /* __FBT__ end */
);

`;

exports[`fbt variable binding detection using ES6 should handle fbt default export 1`] = `
import fbt from "fbt";

fbt._(
  /* __FBT__ start */ {
    type: "text",
    jsfbt: "Foo",
    desc: "Bar",
    project: "",
  } /* __FBT__ end */
);

`;

exports[`fbt variable binding detection using ES6 should handle the named fbt export 1`] = `
import { fbt } from "fbt";

fbt._(
  /* __FBT__ start */ {
    type: "text",
    jsfbt: "Foo",
    desc: "Bar",
    project: "",
  } /* __FBT__ end */
);

`;

exports[`fbt() API:  using FBT subject should accept "subject" as a parameter 1`] = `
const fbt = require("fbt");

fbt._(
  /* __FBT__ start */ {
    type: "table",
    jsfbt: { t: { "*": "Foo" }, m: [{ token: "__subject__", type: 1 }] },
    desc: "Bar",
    project: "",
  } /* __FBT__ end */,
  [fbt._subject(foo)]
);

`;

exports[`fbt() API:  using FBT subject with string templates should accept "subject" as a parameter 1`] = `
const fbt = require("fbt");

fbt._(
  /* __FBT__ start */ {
    type: "table",
    jsfbt: { t: { "*": "Foo" }, m: [{ token: "__subject__", type: 1 }] },
    desc: "Bar",
    project: "",
  } /* __FBT__ end */,
  [fbt._subject(foo)]
);

`;

exports[`fbt() API:  using extraOptions should accept "locale" extra option 1`] = `
const fbt = require("fbt");

fbt._(
  /* __FBT__ start */ {
    type: "text",
    jsfbt: "Foo",
    desc: "Bar",
    project: "",
  } /* __FBT__ end */
);

`;
